### PR TITLE
Add JavaDoc comments to flytekit-api module

### PR DESCRIPTION
**Pull request by pablocasares**
Add JavaDoc comments to flytekit-api module following the protos in flyteidl-protos  module

## Type
 - [x] Documentation
 - [ ] Bug Fix
 - [ ] Feature
 - [ ] Plugin

### CODE REVIEW

File: flytekit-api/src/main/java/org/flyte/api/v1/DynamicJobSpec.java [hunk 1]
```diff
@@ -39,14 +42,14 @@ public abstract class DynamicJobSpec {
   public abstract List<Binding> outputs();
 
   /**
-   * Returns sub-workflows templates.
+   * [Optional] A complete list of task specs referenced in nodes.
```
Comment (narape) [inline flytekit-api/src/main/java/org/flyte/api/v1/DynamicJobSpec.java:45]: Can you check here, this doesnät looks like it belongs to subworkflows

Comment (pablocasares) [reply flytekit-api/src/main/java/org/flyte/api/v1/DynamicJobSpec.java:45]: Hi @narape, I checked both the proto from flyteidl-protos and the original proto from [here](https://github.com/flyteorg/flyteidl/blob/master/gen/pb-go/flyteidl/core/dynamic_job.pb.go#L38) and it seems that the comment in flytekit-java is following the one in flyteidl repo. Do you think we should change it to something different?

Comment (narape) [reply flytekit-api/src/main/java/org/flyte/api/v1/DynamicJobSpec.java:45]: It looks like a copy paste error (in the proto) of the field bellow. Keep the comment that it was before

File: flytekit-api/src/main/java/org/flyte/api/v1/Container.java [hunk 2]
```diff
@@ -24,17 +24,32 @@
...
   public abstract String image();
 
+  /** Environment variables will be set as the container is starting up. */
   public abstract List<KeyValuePair> env();
 
+  /** Container resources requirement as specified by the container engine. */
   @Nullable
   public abstract Resources resources();
 
+  // TODO: add ports and architecture from src/main/proto/flyteidl/core/tasks.proto
```
Comment (lupasarin) [inline flytekit-api/src/main/java/org/flyte/api/v1/Container.java:51]: I guess to be resolved in a separate PR? Just checking if it's something we forgot

File: flytekit-api/src/main/java/org/flyte/api/v1/ContainerError.java [hunk 3]
```diff
@@ -16,11 +16,14 @@
  */
 package org.flyte.api.v1;
 
-/** Exception to propagate detailed errors from container to the execution. */
+/** Error message to propagate detailed errors from container executions to the execution engine. */
 public class ContainerError extends RuntimeException {
 
   private static final long serialVersionUID = 5162780469952221158L;
+  /** A simplified code for errors, so that we can provide a glossary of all possible errors. */
```
Comment (lupasarin) [inline flytekit-api/src/main/java/org/flyte/api/v1/ContainerError.java:23]: Are there known strings that we know the full list of them? If so, can we add a link to that list?

File: flytekit-api/src/main/java/org/flyte/api/v1/ContainerError.java [hunk 4]
```diff
@@ -16,11 +16,14 @@
...
   private final String code;
+  // TODO: add message from src/main/proto/flyteidl/core/errors.proto
```
Comment (lupasarin) [inline flytekit-api/src/main/java/org/flyte/api/v1/ContainerError.java:25]: Is this TODO to be solved here?

File: flytekit-api/src/main/java/org/flyte/api/v1/LaunchPlanIdentifier.java [hunk 5]
```diff
@@ -28,13 +28,19 @@ public static Builder builder() {
 
   @AutoValue.Builder
   public abstract static class Builder {
-
+    /**
+     * Name of the domain the resource belongs to. A domain can be considered as a subset within a
```
Comment (lupasarin) [inline flytekit-api/src/main/java/org/flyte/api/v1/LaunchPlanIdentifier.java:32]:
```suggestion
     * Name of the domain the launch plan belongs to. A domain can be considered as a subset within a
```
(same below)

File: flytekit-api/src/main/java/org/flyte/api/v1/NamedEntityIdentifier.java [hunk 6]
```diff
@@ -19,15 +19,26 @@
...
+   * specific project.
+   */
   public abstract String domain();
 
+  /** Name of the project the resource belongs to. */
   public abstract String project();
 
+  /**
+   * User provided value for the resource. The combination of project + domain + name uniquely
+   * identifies the resource. [Optional] - in certain contexts - like 'List API', 'Launch plans'
```
Comment (lupasarin) [inline flytekit-api/src/main/java/org/flyte/api/v1/NamedEntityIdentifier.java:40]:
```suggestion
   * identifies the resource. [Optional] - used in certain contexts - like 'List API', 'Launch plans'
```

**Review by narape — APPROVED** [review]

File: flytekit-api/src/main/java/org/flyte/api/v1/TaskTemplate.java [hunk 7]
```diff
@@ -29,19 +29,39 @@ public abstract class TaskTemplate {
...
+  /**
+   * A strongly typed interface for the task. This enables others to use this task within a workflow
+   * and guarantees compile-time validation of the workflow to avoid costly runtime failures.
+   */
   public abstract TypedInterface interface_();
 
+  /**
+   * A predefined yet extensible Task type identifier. This can be used to customize any of the
+   * components. If no extensions are provided in the system, Flyte will resolve this task to its
+   * TaskCategory and default the implementation registered for the TaskCategory.
```
Comment (lupasarin) [inline flytekit-api/src/main/java/org/flyte/api/v1/TaskTemplate.java:41]: I don't fully understand this part: "If no extensions are provided in the system, Flyte will resolve this task to its
   * TaskCategory and default the implementation registered for the TaskCategory."

File: flytekit-api/src/main/java/org/flyte/api/v1/Variable.java [hunk 8]
```diff
@@ -23,8 +23,10 @@
 @AutoValue
 public abstract class Variable {
 
+  /** Variable literal type. */
   public abstract LiteralType literalType();
 
+  /** [Optional] string describing input variable. */
```
Comment (lupasarin) [inline flytekit-api/src/main/java/org/flyte/api/v1/Variable.java:29]: Is Variable always used to refer to an input?

File: flytekit-api/src/main/java/org/flyte/api/v1/Container.java (hunk 2 above)
Comment (pablocasares) [reply flytekit-api/src/main/java/org/flyte/api/v1/Container.java:51]: The added TODO's should not be resolved in this PR. They were added to track missing fields in the java classes from the proto files.
